make wxScrolled<> autoscroll region configurable

wsu-cb opened the pull request:
This is an initial attempt at configuring `wxScrolled<>`'s autoscroll region.  The default configuration emulates the existing behavior, but can be overridden to change the autoscroll region.  In particular, my personal requirement is for the region to be inside the window, but near the edge, but the configuration functions allow for some other possibilities.

Would anyone prefer a different API?  Or have any other comments?

vadz reviewed (commented):
Sorry, I'm really confused by the API, so I'd appreciate some clarification/examples. TIA!

wsu-cb committed:
Improve auto-scrolling in wxScrolled

Support auto-scrolling both orientations simultaneously and make
auto-scroll region configurable.

Also allow changing auto-scroll dir without ending drag operation and
starting another one.

Closes #25978.

wsu-cb commented:
> If the autoscroll code is touched at all, i think it should also fix the (IMHO) only flaw it currently has: It can't scroll in both directions at the same time. E.g. if you drag and move the mouse outside at the bottom, it starts scrolling down. But if you then also move the mouse outside to the right side, it won't start also scrolling in that directions.

I have made this change in https://github.com/wxWidgets/wxWidgets/pull/25978/commits/81103a58cd078417fd6780e1d85cb244d5e3198f, and then re-built the autoscroll zone changes on top of that.

vadz commented on include/wx/scrolwin.h:
```diff
@@ -345,6 +364,10 @@ class WXDLLIMPEXP_CORE wxScrollHelperBase : public wxAnyScrollHelperBase
 
+    wxCoord               m_innerScrollZone = wxDefaultCoord;
+    wxCoord               m_outerScrollZone = wxDefaultCoord;
+    bool                  m_inNonScrollRegion = true;
```
Why use negation here? Wouldn't


```suggestion
    bool                  m_inAutoScrollRegion = false;
```

be clearer?

vadz reviewed (commented):
Sorry for the long delay, I've finally looked at this again and I think I understand the logic better now, so I'm ready to merge this, but I wonder if this could be improved a bit first. Notably I'd really like to avoid using `std::pair` and could make the proposed change myself if you agree.

I think trying to make things more clear is worth it as this is not the most obvious part of code already...

vadz reviewed (approved):
Looks good, thanks! Will merge soon.

wsu-cb commented:
> Looks good, thanks! Will merge soon.

I'm certainly not going to complain if you merge now, but, now that you have approved the implementation,I was planning to document `SetInnerScrollZone()` and `SetInnerScrollZone()` in `interface/wx/scrolwin.h` before merging.

vadz committed:
Document wxScrolled::Set{Inner,Outer}ScrollZone()

vadz commented:
> I prefer that the default value be `wxDefaultCoord`, but, as maintainer, you have the final decision.

I'd rather reach some consensus. I understand that it's better to use `wxDefaultCoord` as default value, but this is in conflict with inner autoscroll being disabled by default and outer being enabled. And IMO it's worse for the same value to mean 2 different things in related contexts.

What do you think about `EnableInnerAutoScroll(width)` and `DisableOuterAutoScroll()` idea? The names could be improved, e.g. `EnableAutoScrollInside()` and `DisableAutoScrollOutside()` might be better.

wsu-cb commented:
> Shouldn't outer zone always cover everything outside of the client area?

I think this is the crux of the issue.  As I wrote in https://github.com/wxWidgets/wxWidgets/pull/25978#discussion_r2595800569, it is entirely possible that I have attempted too much generality.  Because I don't like the outer scroll zone at all, I am unqualified to determine whether a partial outer scroll zone is ever useful.

> This is not objectionable, of course. It's the difference between interpretations of `-1` between zones which is: for the inner zone it is (roughly) the same as `0`, but for the outer zone it's (roughly) exactly the opposite of `0`.

I see your point now, and can't dispute it.

For my purposes, the suggested API in https://github.com/wxWidgets/wxWidgets/pull/25978#issuecomment-3902554264 would work well.  I chose a general API to avoid objections about it being inconsistent to permit a partial inner scroll zone but not a partial outer scroll zone, but if no one values that generality, I actually prefer the simpler API as well.

wsu-cb commented:
> I will look at your branch and get back to you...

After experimenting, I now think the additional complexity in my previous design of sometimes setting the autoscroll region relative to the client rectangle is also not worthwhile.  I now think I will revert to the wxWidgets 3.2 behavior of always basing the autoscroll region on the window rectangle (although the implementation is unrelated to the 3.2 implementation using `wxEVT_ENTER_WINDOW`/`wxEVT_LEAVE_WINDOW`).

wsu-cb committed:
wxScrolled<>:  simplify autoscroll configuration API

wsu-cb commented:
> Yes, thank you, this looks good, the only remaining question is whether we should allow passing `wxDefaultCoord` to `EnableAutoScrollInside()`. As long as it works in the same way as `0`, I would say that we shouldn't and that the function should assert that its argument is `>= 0`. And setting it to 0 would disable inner auto scroll region, of course.

As noted in https://github.com/wxWidgets/wxWidgets/pull/25978#issuecomment-3902538102, my preference is that the default value be `wxDefaultCoord`, even if it acts the same as `0`, and even if it means a little bit of extra overhead in the implementation.

vadz commented:
> it seems perfectly reasonable to me for `wxDefaultCoord` to mean something that can also be specified explicitly

I can't think of any occurrence of this in wxWidgets API. `wxDefaultXXX` always means "unspecified" or, well, "default". There is no reasonable default value for the inner auto scroll zone size, but I'd actually expect setting it to `wxDefaultCoord` to do exactly this, i.e. choose some predefined value like `FromDIP(5)`, rather than disable it entirely.

In fact, you've helped me put my finger on it: this API looks wrong to me because it uses `wxDefaultCoord` as meaning "none" and this is just not how it's used anywhere else.

wsu-cb commented:
> I can't think of any occurrence of this in wxWidgets API. `wxDefaultXXX` always means "unspecified" or, well, "default". There is no reasonable default value for the inner auto scroll zone size, but I'd actually expect setting it to `wxDefaultCoord` to do exactly this, i.e. choose some predefined value like `FromDIP(5)`, rather than disable it entirely.

This appears to be where we disagree:  I think `0` is the default value.  The difference between `1` and `0` is enormous in a logical sense, but for the user, a 1-pixel wide scroll zone is almost as unhittable as a 0-pixel wide scroll zone.

> 
> In fact, you've helped me put my finger on it: this API looks wrong to me because it uses `wxDefaultCoord` as meaning "none" and this is just not how it's used anywhere else.

wsu-cb committed:
wxScrolled<>:  simplify autoscroll zone config API

wsu-cb commented:
> One way or another, this needs to be finalized soon if it's to be included in 3.3.2.

I have pushed a new version with the simplified API discussed above.  (This implementation forbids `wxDefaultCoord`.)

vadz commented:
Thanks, I'll merge and push this soon, it can always be improved/extended later if necessary.
